### ISSUE TITLE
Separate type specs from predicates

### ISSUE THREAD

**Issue by solnic** [dry-rb/dry-validation#181]
Applying basic type-checking predicates should be enforced by the DSL, otherwise we may have schemas with pretty much broken rules leading to issues like #145 or #176 

To improve the situation we should have a new API that's dedicated for defining a type expectation and everything else will be built on top of that. My proposition:

``` ruby
Dry::Validation.Form do
  # => name must be present and it's a non-empty string (where :str type is default in Form)
  required(:name).filled

  # => age must be present and it's either an int or a nil
  required(:age, [:nil, :int]).maybe

  # => tags must be present and it's either an empty array or a 2-element
  #    array with string values
  required(:tags, :array) { empty? | size?(2) & each(:str?) }

  # => labels must be present and it's a non-empty array or string
  required(:labels, [:array, :str]).filled
end
```

closes #145 
closes #174 
closes #176 
closes #182 
closes #188

**Comment by fran-worley** [comment]
For what it's worth, I like it. 

Worth noting that this has implications for Reform as your effectively to be declaring types twice. /c @apotonick

**Comment by mensfeld** [comment]
I think this might be related:
```
key(:stdout) { str? | array? | hash? }
`instance_eval': wrong number of arguments (given 0, expected 1..3) (ArgumentError)

key(:stderr) { str? | array? | type?(Hash) } #=> this works
```

**Comment by solnic** [comment]
@mensfeld that's a bug in `hash?` which has special handling in the DSL and it's a known issue that will be fixed, both syntax should work the same

**Comment by jodosha** [comment]
@solnic With this new proposal, would the following cases still work?

```ruby
required(:name).filled # Omits type
```

```ruby
required(:name).filled(:str?) # Old syntax
```

**Comment by solnic** [comment]
@jodosha yes but both would give you a deprecation warning, the idea is to make type expectations more explicit

**Comment by jodosha** [comment]
@solnic Isn't the design odd? I mean, we will need two method invocations to make it to work properly. There isn't any hint that I need to call `type` after `required`.

For instance a, single method with two arguments guides better developers.

```ruby
required(:name, :str?)
```

If a developer uses with only `:name` they will get an error. wdyt?

**Comment by solnic** [comment]
I agree with you. Having second thoughts about this and your suggestion
makes perfect sense. In fact, it may simplify other things too.

**Comment by solnic** [comment]
OK so this turned out to be way more challenging than I thought it would be, I didn't think this through when proposing this API. Here are a couple of things that should be taken into consideration when designing this interface:

- A value can have multiple types, which means we may have additional rules depending on the type like `str? & size?(1..10) | array? & size?(3)` and *this* makes it tricky
- Our `maybe` macro (which covers a very common usecase) no longer has its place when the type spec is separated, so *probably* it would have to be deprecated
- Specifying multiple types as a list wouldn't make it clear for which types further rules are being applied, and we'd end up with implicit behavior again

So, this is tough, and now I remember why I've created coercion inference based on rule tree, as it makes it possible to create many rules for many types and you still get proper coercions, this is a *really advanced* functionality and I feel like this should stay but we somehow should make sure types are specified. Furthermore, removing complex input processor compilers should be possible...we can capture type specs via predicate names and build a type-map inside the DSL, then if there are missing type specs we could raise an error for schemas with input-processor configured (as it *needs* type specs).

Bottom line - I think we shouldn't introduce this explicit type-spec interface after all, just because it cannot be flexible enough, instead I'll try to simplify how input-processor is being created and make sure we can notify the user that some rules are missing type specs.

**Comment by solnic** [comment]
OK 2 hours later - this is too complex, especially with nested schemas, so how about introducing an explicit API like that:

``` ruby
# using type identifiers compatible with dry-types
required(:email, :str)
required(:age, [:nil, :int])
required(:address).schema { required(:street, :str) }

# this would result in having a hash schema from dry-types like this:
Types::Form::Hash.weak(
  email: Types::String,
  age: Types::Form::Nil | Types::Form::Int,
  address: Types::Form.weak(street: Types::String)
)
```

This would **completely separate** type specs from predicates, and we could deprecate complex coercion inference too, and remove it in 1.0.0. There are other benefits, like the fact you could pass in your own coercion object (could be anything responding to `#[]`).

This pretty much means that we're going back to the idea that coercion should be completely separated from validation, my past-self and @AMHOL are laughing at me right now ;)

**Comment by fran-worley** [comment]
@solnic what are the implications of this with the discussion that we were having in Reform (https://github.com/apotonick/reform/pull/347)? /cc @apotonick

Also, how does you revised API work with the maybe macro?
```
required(:age).maybe(:int)  
#becomes:
required(:age, [:nil, :int])
#???
```

**Comment by solnic** [comment]
@fran-worley the implication is huge - if you don't define types in `required` or `optional` then you won't get coercions (eventually, in 1.0.0, now I'd just keep original behavior but add a warning when you don't define types explicitly)

**Comment by fran-worley** [comment]
@solnic I meant more the divergence of opinion between Reform which handles types via Property and dry-v which currently handles them during validation.

Would you be able to separate the two entirely? So Reform could build the hash schema via the property call and then append any validations ontop?

**Comment by jodosha** [comment]
@solnic I don't mind the last API proposal, but how does it works with multiple types? Eg. the example you posted above `str? & size?(1..10) | array? & size?(3)`.

To make this more visually pleasant, can we consider to:

  * Allow both symbols and constants for type definitions? Eg. `:int` and `Integer` are equivalent. This is for consistency with other libraries like ROM migrations and mapping, or Dry Types stucts.

  * Make the signature to accept splat arguments: `required(:age, :nil, :int)`.

**Comment by solnic** [comment]
@fran-worley yeah reform can easily do that, it could use meta-data from schema definition to build its property stuff so that people won't have to define same stuff using two APIs

@jodosha predicates stay the same, you'd define valid type(s) separately from rules, so things will work the same way as now wrt rules

> Allow both symbols and constants for type definitions? Eg. `:int` and `Integer` are equivalent. This is for consistency with other libraries like ROM migrations and mapping, or Dry Types stucts.

We could accept primitive classes that would be mapped to specific dry-types, sure, we should also support dry-types objects so `Types::Int` etc. Personally I'd probably just use symbols as they are shorter :)

> Make the signature to accept splat arguments: required(:age, :nil, :int).

The reason why I used an array was to make it clear that the first arg is the name and the second is for type def, not sure which is better though.

**Comment by timriley** [comment]
I like your most recent proposal – I think the explicit type declaration will make schemas easier to write and understand, and the possibility of passing your own typecasting objects (responding to `#[]`) is a nice bonus. The inferred coercion was advanced and it would've been difficult for people to hook into it themselves.

> The reason why I used an array was to make it clear that the first arg is the name and the second is for type def, not sure which is better though.

Yeah. I think `required(:age, [:nil, :int])` reads more clearly to me.

**Comment by fran-worley** [comment]
@solnic I think that this will be a big plus for Reform then. Are you planning this for 0.8 or 1.0? 

How does it work with validations conditional on a type? 

e.g. an input could be a string or an array. If an array do X if a string to Y.

**Comment by solnic** [comment]
@fran-worley this will be done for 0.8.0 - if we can agree that what I proposed makes sense, I'll Just Do It™ and it shouldn't be hard (famous last words!)

> How does it work with validations conditional on a type?

Predicates will work *the same*, no change is needed there, we're talking about *separating type definitions from predicates completely*, which means we will no longer infer types from predicates in 1.0.0.

**Comment by fran-worley** [comment]
@solnic that's great then. For what its worth, I agree that your proposal makes sense 👍

**Comment by solnic** [comment]
I updated the description of this issue with the new proposal. I also realized we *could* infer *predicates* from *type specs*, doing that in this direction will be very simple.

**Comment by solnic** [comment]
Please check out the PR which implements this #189
